samplesheets/assayapps/generic/plugins.py

```diff
@@ -0,0 +1,175 @@
+    # TODO: TBD: Do we need this?
+    permission = None
+
+    #: Toggle displaying of row-based iRODS links in the assay table
+    display_row_links = True
+
+    @classmethod
+    def _get_col_value(cls, colname, row, table):
+        """
+        Return value of last matched column.
+
+        :param colname: Column name to look for
+        :param row: List of dicts (a row returned by SampleSheetTableBuilder)
+        :param table: Full table with headers (dict returned by
+                      SampleSheetTableBuilder)
+        :return: String with cell value of last matched column
+        """
+        # returns last match of row
+        value = ''
+        for i in range(len(row)):
+            header = table['field_header'][i]
+            if header['value'].lower() == colname.lower():
+                value = row[i]['value']
+        return value
+
+    def get_row_path(self, row, table, assay, assay_path):
+        """
+        Return iRODS path for an assay row in a sample sheet. If None,
+        display default path. Used if display_row_links = True.
+
+        :param row: List of dicts (a row returned by SampleSheetTableBuilder)
+        :param table: Full table with headers (dict returned by
+                      SampleSheetTableBuilder)
+        :param assay: Assay object
+        :param assay_path: Root path for assay
+        :return: String with full iRODS path or None
+        """
+        data_cols = []
+        for c in DATA_COMMENTS:
+            data_col = assay.comments.get(c)
+            data_cols.append(self._get_col_value(data_col, row, table))
+
+        # Build iRODS path from list and stop at first None value
+        data_path = ''
+        for c in data_cols:
+            if c:
+                data_path += '/' + c
+            else:
+                break
+
+        if data_path:
+            return assay_path + data_path
+        return None
+
+    def update_row(self, row, table, assay):
+        """
+        Update render table row with e.g. links. Return the modified row.
+
+        :param row: Original row (list of dicts)
+        :param table: Full table (dict)
+        :param assay: Assay object
+        :return: List of dicts
+        """
+        if not settings.IRODS_WEBDAV_ENABLED or not assay:
+            return row
+        assay_path = self.get_assay_path(assay)
+        if not assay_path:
+            return row
+
+        base_url = settings.IRODS_WEBDAV_URL + assay_path
+        top_header = None
+        th_colspan = 0
+
+        for i in range(len(row)):
+            header = table['field_header'][i]
+            if not top_header or i >= th_colspan:
+                top_header = get_top_header(table, i)
+                th_colspan += top_header['colspan']
+
+            # TODO: Check if two comments reference the same column header?
+            # Create Results & Reports links
+            if assay.comments.get(RESULTS_REPORTS_COMMENT):
```

Comment:
There seems to be repeated code for the results and reports collections and the misc files collections. Maybe make a single method out of that and call it with the comment as an arg.

Reply:
implemented, it's a bit messy though